### PR TITLE
Add support for scraping FotMob

### PR DESCRIPTION
**Pull request by marcjbaron**
Hello, 

This is an attempt at adding Fotmob as a data source to soccerdata. The nox test suite passed without errors, with the exception of the python-3.8 tests (but the current python dependency in the pyproject.toml file is >=3.9, so I don't know if this can be safely ignored). Unit tests were also created and passed successfully. 

The current methods are:
 - read_seasons(), read_leagues(), read_schedule() (similar to the other data sources)
 - read_match_stats() -> reads stats for an individual game
 - read_league_table() -> this functionality is not included in the other data sources, but I needed it for a personal project, so kept it in.

There are obviously other methods that could be added (e.g. season stats), but did not want to go much further until early issues are dealt with.

### CODE REVIEW

File: soccerdata/fotmob.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        """Initialize the Fotmob reader."""
+        super().__init__(
+            leagues=leagues,
+            proxy=proxy,
+            no_cache=no_cache,
+            no_store=no_store,
+            data_dir=data_dir,
+        )
+        self.seasons = seasons  # type: ignore
+        self.rate_limit = 3
```
Comment (probberechts) [inline soccerdata/fotmob.py:79]: This is a very high rate limit. Does FotMob implement some kind of countermeasure against scraping?

File: soccerdata/fotmob.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        self.max_delay = 3
+        if not self.no_store:
+            (self.data_dir / 'seasons').mkdir(parents=True, exist_ok=True)
+            (self.data_dir / 'matches').mkdir(parents=True, exist_ok=True)
+            (self.data_dir / 'previews').mkdir(parents=True, exist_ok=True)
+            (self.data_dir / 'events').mkdir(parents=True, exist_ok=True)
```
Comment (probberechts) [inline soccerdata/fotmob.py:85]: It seems you are not using any of these directories.

File: soccerdata/fotmob.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        return res
+
+    def read_leagues(self) -> pd.DataFrame:
+        """Retrieve the selected leagues from the datasource.
+
+        Returns
+        -------
+        pd.DataFrame
+        """
+        filemask = 'allLeagues'
```
Comment (probberechts) [inline soccerdata/fotmob.py:105]: This would be a urlmask.

File: soccerdata/fotmob.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        url = FOTMOB_API + filemask
+        filepath = self.data_dir / 'allLeagues.json'
+        reader = self.get(url, filepath)
+        data = json.load(reader)
+        leagues = []
+        for k, v in data.items():
+            if (k == 'favourite') | (k == 'popular') | (k == 'userSettings'):
```
Comment (probberechts) [inline soccerdata/fotmob.py:112]: Use "or" instead of "|". Or even better `k in ("favourite", ...)`

File: soccerdata/fotmob.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        Returns
+        -------
+        pd.DataFrame
+        """
+        df_leagues = self.read_leagues()
+        seasons = []
+        for lkey, league in df_leagues.iterrows():
+            url_append = 'leagues?id=' + str(league.league_id)
+            url = FOTMOB_API + url_append
+            print(url)
```
Comment (probberechts) [inline soccerdata/fotmob.py:157]: Can you make remove all print statements?

File: soccerdata/fotmob.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+            # stats.append(tmp_dfs)
+            df_game = (
+                pd.concat(stats, axis=1).replace({'team': TEAMNAME_REPLACEMENTS}).sort_index()
+            )
+            # df_game = df_game.T.drop_duplicates().T
+            games.append(df_game)
+        df = (
+            pd.concat(games, ignore_index=True)
+            .replace({'team': TEAMNAME_REPLACEMENTS})
+            .set_index(['Matchweek', 'Date', 'Home Team', 'Home id', 'Away Team', 'Away id'])
```
Comment (probberechts) [inline soccerdata/fotmob.py:431]: Indexes should be consistent across data sources using the following keys: ["league", "season", "game", "team", "player"]

File: soccerdata/fotmob.py (hunk 1 above)
Comment (marcjbaron) [reply soccerdata/fotmob.py:79]: Not that I'm aware of; I just cautiously set the same rate limit to be the same as fbref, but can see what happens when removed.

**Comment by marcjbaron** [comment]
I have made the requested changes (and a few additional changes for clarity). 

If/when the PR is approved, I also have a "read_team_season_stats" method ready.

**Comment by probberechts** [comment]
Something seems to be wrong with the `poetry.lock` file. Which version of poetry are you currently using? I am on 1.7.1. Also, can you restore the lock file to the original one? Feature PR's should not change the versions of dependencies.

**Comment by marcjbaron** [comment]
Lock file should now match the original.

**Comment by probberechts** [comment]
Thanks. I did some refactoring, fixed a few more bugs and simplified the columns in the output dataframes. The only thing that still does not seem to work is retrieving the table for tournaments. Currently, it gets the ranking of a single group only.

**Comment by marcjbaron** [comment]
Thanks for cleaning up the code. 

The tournament tables are potentially fixed: I added a column called "group" to the each table; for leagues without a group stage, the column will return 'NaN'. 

On that note, for previously completed seasons, FotMob's api gives 'null' for the round and week in the fixtures list, so these are returned as 'None' in the read_schedule method.

I'm not sure if either of these are desirable though.

**Review by probberechts — APPROVED** [review]

**Pull request merged**